Preserve current `Animation` progress if `SpeedRatio` or `PlaybackDirection` changes

timiimit-tam opened the pull request:
Apologies for not opening an Issue first... Let me know if there's a better way of doing this and I can change anything.

## What does the pull request do?
This PR makes it possible to change `SpeedRatio` and `PlaybackDirection` while animation is running without affecting the current animation progress. PR keeps compatibility with old behavior when speed does not change.


## What is the current behavior?
When `Animation.SpeedRatio` or `Animation.PlaybackDirection` changes, the animation progress/timeline **IS** affected.

## What is the updated/expected behavior with this PR?
When `Animation.SpeedRatio` or `Animation.PlaybackDirection` changes, the animation progress/timeline **IS NOT** affected.

## How was the solution implemented (if it's not obvious)?

### In short
When relevant properties change during a running animation, the animation's current progress is remembered and animation continues from that point on with respect to property changes.

This is all that's needed to handle a changing `SpeedRatio`. For `PlaybackDirection`, a lot more care has to be put into the logic as explained in the following subsections.

### Definition of proper animation direction
*Proper animation direction* is the value of `PlaybackDirection` at first frame of the animation. In code it is indicated by `_timeMovesBackwards` field. `PlaybackDirection.Normal` and `PlaybackDirection.Alternate` result in positive *proper animation direction*, while `PlaybackDirection.Reverse` and `PlaybackDirection.AlternateReverse` result in negative *proper animation direction*. If `PlaybackDirection` changes direction some time after the first frame of animation, the animation will start playing backwards. If left running, animation will eventually reach the starting position of animation the second time.

### Behavior at starting position
When animation begins playing there is nothing new to discuss, it plays as it always has. When animation reverses direction and reaches the starting position subsequently, animation will **always** hold value of initial keyframe (first or last keyframe depending on *proper animation direction*), no matter the value of `FillMode`. 

Animations with INFINITE `IterationCount` will continue backwards past the starting position of the animation. This is to allow changing animation direction without affecting the infinitely repeating nature of the animation.

Animations with finite `IterationCount` will be clamped to the starting position and will wait for direction to change back into *proper animation direction*. When that happens, animation will again start playing normally the same way as it has when animation started the first time (albeit with a potentially different speed). Test [Reversing_Direction_Past_Initial_Point_Clamps_To_Initial_Point](https://github.com/AvaloniaUI/Avalonia/pull/19830/commits/343e6b49ab3189b8ea1e90994bf3872a6a327ebd#diff-2951659089d45c880a0b434f05676802e2886a24039a5847559a46ee0803a036R773) has been added to check the correctnes of this behavior.

### Behavior of delays
Between the starting position and ending position of animation, behavior of both `Delay` and `DelayBetweenIterations` is always the same, no matter how many times any given delay has been reached. This can be though of as delays being "*baked-into*" the animation. I believe this is what most people would expect and want.

The only new delay behavior is with animations where `IterationCount` is INFINITE. When animation plays backwards past the starting point, the `DelayBetweenIterations` will be played in the opposite part of the iteration time. Here is a quick/crudely drawn illustration of animation timeline to make it clearer:
<img width="655" height="195" alt="image" src="https://github.com/user-attachments/assets/87c47269-809d-4b3a-8284-e6aad534d438" />
If user wants to always have the same delay between all iterations, then both `Delay` and `DelayBetweenIterations` need to be set to the same value.

### Unhandled case
Changing between `Alternate` and non-`Alternate` `PlaybackDirection`s are not handled. This means that switching between such two modes (from `Normal` or `Reverse` to `Alternate` or `AlternateReverse` or vice versa) can result in a "jump" of current animation progress depending on when you switch it.

This could be handled in a way that kind of "does" make sense, but I don't see a reason why anyone would need to switch between these two modes.

### Finally
Please review carefully in case I missed something, I wouldn't want to break animations which are probably considered a core feature.

I should also mention that I didn't do any testing of `FillMode` (besides what I mentioned above), I just assumed that if all UnitTests pass, then it works as expected.

## Checklist

- [x] Added `Animation Speed` page to RenderDemo sample for preview
- [x] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

## Breaking changes
`IterationCount.Value` may no longer be greater than `long.MaxValue`.

## Obsoletions / Deprecations
NONE

## Fixed issues
Fixes #19979

MrJul committed:
Merge branch 'master' into anim-speedratio-fix

MrJul commented on src/Avalonia.Base/Animation/AnimationInstance`1.cs:
```diff
@@ -109,121 +135,287 @@ public void Step(TimeSpan frameTick)
+            }
+        }
+
+        private bool IsIterationReversed(bool isAlternatingPlaybackDirection, long? iterationIndex)
```
Nit: can be made static

MrJul commented on src/Avalonia.Base/Animation/AnimationInstance`1.cs:
```diff
@@ -109,121 +135,287 @@ public void Step(TimeSpan frameTick)
+            }
 
-                // Check if normalized time needs to be reversed according to PlaybackDirection
+            var timeUntilIterDelay = iterDuration - iterTime;
```
Unused variables, leftover?

MrJul commented on src/Avalonia.Base/Animation/AnimationInstance`1.cs:
```diff
@@ -109,121 +135,287 @@ public void Step(TimeSpan frameTick)
+        }
 
-            if (!_gotFirstKFValue)
+        private void ApplyLimitedClamp(ref long animTime, long animTimeOrigin)
```
Unused parameter `animTimeOrigin`

MrJul reviewed (commented):
Thank you for your contribution!
The PR is very well described and the code clean and easy to follow.

That said, please bear with me while I review this, it's been a while since I dove into the animation code and while well written, the animation algorithm is starting to be quite complex :)

timiimit-tam commented on src/Avalonia.Base/Animation/AnimationInstance`1.cs:
```diff
@@ -63,25 +79,35 @@ private void FetchProperties()
-                _iterationCount = _animation.IterationCount.Value;
+            {
+                if (_animation.IterationCount.Value > long.MaxValue)
+                    throw new InvalidOperationException("IterationCount value cannot be larger than long.MaxValue.");
```
This `if` statment can also include a check that would prevent algo breakage. And I see 2 solutions for that. Change condition to check for `long.MaxValue / iterDurationTotal` and:
- Throw if its higher
- Set `_iterationCount = null`

This also raises another issue with the code. For which I'm unsure whether I should also try to handle or if im just being too pedantic. `IterationCount.INFINITE` will break the animation logic same way as another really high value does. Should I make sure to explicitly handle `animTime` over/under-flow? I'm pretty sure this issue has been present already in old animation logic. Maybe it would be smart to just skip unneccessary computation for now unless someone *actually* needs it and reports an issue.

timiimit-tam committed:
Add AnimationSpeedPage to RenderDemo

timiimit-tam committed:
Implement fix

timiimit-tam committed:
Add tests

timiimit-tam committed:
Minor improvements

Remove leftover vars/params
Replace some `if`s with terniary assignment
Improve comments
Unduplicate `playbackReversed` code

timiimit-tam committed:
Add missing check for negative `DelayBetweenIterations`

timiimit-tam committed:
Add animation time precision

timiimit-tam committed:
Add unit test

DelayBetweenIterations_Behind_Initial_Point_Is_In_Front_Of_Iterations

timiimit-tam committed:
Merge branch 'AvaloniaUI:master' into anim-speedratio-fix

timiimit-tam committed:
Merge branch 'master' into anim-speedratio-fix

MrJul committed:
Merge branch 'main' into fork/timiimit-tam/anim-speedratio-fix

# Conflicts:
#	src/Avalonia.Base/Animation/AnimationInstance`1.cs

MrJul committed:
Fix zero duration regression
